java/jsinterop/generator/visitor/UnionTypeHelperTypeCreator.java

```diff
@@ -80,12 +80,18 @@ public void applyTo(Program program) {
 
           @Override
           public boolean shouldProcessField(Field field) {
+            if (field.isStatic()) {
+              currentNameStack.push("Static");
+            }
```

Comment:
I need to think more about that. This is a breaking change. every static fields/methods with an union type or funtcion type will see a change in the name of their UnionType helper.

Reply:
The case I'm specifically working on here is that the protobuf impl for JS offers instance and static methods with the same names, and the same overloads, so the exact same union gets generated twice. In a specific case like that, it could even be possible to observe that the union and the method name are the same, so share the union type rather than generate it again.

Reply:
I changed this code a bit - it isn't pretty, but hoping it touches the basics of what you were thinking of? In short, if the field is static, and there is another field with the same name, then we need the `Static` suffix - naturally this won't affect any existing code, since it would have been affected by this bug.

For methods it is slightly more complicated, we need to only test the non-overlay methods, since those just call the one "real" method, which takes the union type.

I havent updated tests yet, but presently all existing and new tests pass - once we settle on the approach I'll get those in line with what you are requesting there too?

Reply:
I think we should take another approach. The MembersClassCleaner is doing too much thing: it's renaming methods/fields, removing constructors/methods or fields and modifying method's generics. We should split that in different visitors: I think at least two in order to modify the method generics in a separate visitor. 
The visitor that will rename class members can be run before the UnionTypeMethodParameterHandler and the one that modify method's generics need to be called after the FixReferencesToDuplicatedTypes visitor.
I think it will solve all the issues.

Reply:
Okay, I'll close this for now - it will probably be a few months until I have something to show here, I'm going on personal leave sometime in the next week and will not be in a position any more to take on work like this.

Reply:
No problem and thanks your patch and opening the issue. I can take over the work.

Reply:
Hi, checking back on this, are you still able to resolve this some other way?